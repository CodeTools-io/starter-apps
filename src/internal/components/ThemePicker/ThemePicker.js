import React, { useMemo, useState } from 'react';
import {
  Box,
  Button,
  Heading,
  Select,
  Text,
  Layer,
  MaskedInput,
} from 'grommet';

const THEME_DESIGNER_ORIGIN = process?.env?.REACT_APP_THEME_DESIGNER_ORIGIN;
const THEME_DESIGNER_STORAGE = process?.env?.REACT_APP_THEME_DESIGNER_STORAGE;

export default function ThemePicker({
  currentTheme,
  onChange = () => {},
  onImport = () => {},
}) {
  const [themeOptions, setThemeOptions] = useState([
    'caribbean',
    'default',
    'paradise',
    'sunglow',
  ]);
  const [inputValue, setInputValue] = useState('');
  const [importingTheme, setImportingTheme] = useState(false);
  const showImport = useMemo(() => {
    return currentTheme === 'import…' && !importingTheme;
  }, [currentTheme, importingTheme]);

  function addTheme(themeName) {
    if (!themeOptions?.includes(themeName)) {
      setThemeOptions([...themeOptions, themeName]);
    }
  }

  function importTheme(url) {
    setImportingTheme(true);

    let themeEndpoint = url;

    if (themeEndpoint.startsWith(THEME_DESIGNER_ORIGIN)) {
      const themeUrl = new URL(themeEndpoint);
      const themeParams = new URLSearchParams(themeUrl.search);
      if (themeParams?.has?.('id')) {
        themeEndpoint = `${THEME_DESIGNER_STORAGE}/${themeParams?.get?.('id')}`;
      }
    }

    fetch(`${themeEndpoint}`)
      .then((res) => {
        if (res?.status !== 200) {
          throw new Error('Could not retrieve the theme');
        }

        return res.json();
      })
      .then((data) => {
        addTheme(data?.name);
        onImport({ theme: data });
        setInputValue('');
      })
      .catch((err) => {
        console.error(err);
      })
      .finally(() => {
        setImportingTheme(false);
      });
  }

  return (
    <Box className="ThemePicker">
      <Text color="text-xweak" size="small" margin={{ top: '-1rem' }}>
        theme
      </Text>
      <Select
        options={[...themeOptions, 'import…']}
        value={currentTheme}
        onChange={onChange}
      />
      {showImport && (
        <Layer position="center">
          <Box pad="medium" gap="small" width="medium">
            <Heading level={3} margin="none">
              Import
            </Heading>
            <Text>What is the url of the theme?</Text>
            <MaskedInput
              mask={[{ fixed: 'https://' }, { regexp: /^.*$/ }]}
              value={inputValue}
              onChange={(event) => setInputValue(event.target.value)}
            />
            <Box
              as="footer"
              gap="small"
              direction="row"
              align="center"
              justify="end"
              pad={{ top: 'medium', bottom: 'small' }}
            >
              <Button
                label="Cancel"
                onClick={() => {
                  onChange({ value: 'default' });
                }}
                color="dark-3"
              />
              <Button
                label={
                  <Text color="white">
                    <strong>Import</strong>
                  </Text>
                }
                onClick={() => importTheme(inputValue)}
                primary
              />
            </Box>
          </Box>
        </Layer>
      )}
    </Box>
  );
}
