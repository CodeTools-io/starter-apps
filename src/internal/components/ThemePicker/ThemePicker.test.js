import React from 'react'
import ThemePicker from './ThemePicker'

describe('ThemePicker', () => {
  test('can render', () => {
    const { container } = renderComponent(
      <ThemePicker>
        <p>Content for the ThemePicker</p>
      </ThemePicker>
    )

    expect(container).toMatchSnapshot()
  })
})

