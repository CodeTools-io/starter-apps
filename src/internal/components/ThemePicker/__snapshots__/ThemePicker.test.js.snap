// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemePicker can render 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 dYebPD ThemePicker"
  >
    <span
      class="StyledText-sc-1sadyjn-0 dXXMHF"
    >
      theme
    </span>
    <button
      aria-label="Open Drop"
      class="StyledButton-sc-323bzc-0 ebqXLG Select__StyledSelectDropButton-sc-17idtfo-1 iGEQF"
      type="button"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 jmuHez"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 iAryDB"
        >
          <div
            class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
          >
            <input
              autocomplete="off"
              class="StyledTextInput-sc-1x30a0s-0 gDaIpF Select__SelectTextInput-sc-17idtfo-0 bIurki"
              readonly=""
              tabindex="-1"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="StyledBox-sc-13pk1d4-0 eA-DBEo"
          style="min-width: auto;"
        >
          <svg
            aria-label="FormDown"
            class="StyledIcon-ofa7kd-0 slgkV"
            viewBox="0 0 24 24"
          >
            <polyline
              fill="none"
              points="18 9 12 15 6 9"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </button>
  </div>
</div>
`;
